src/jit/importer.cpp

```diff
@@ -18393,4 +18399,366 @@ bool Compiler::IsMathIntrinsic(GenTreePtr tree)
 {
     return (tree->OperGet() == GT_INTRINSIC) && IsMathIntrinsic(tree->gtIntrinsic.gtIntrinsicId);
 }
-/*****************************************************************************/
+
+//------------------------------------------------------------------------
+// impDevirtualizeCall: Attempt to change a virtual vtable call into a
+//   normal call
+//
+// Arguments:
+//     call -- the call node to examine/modify
+//     thisObj  -- the value of 'this' for the call
+//     callInfo -- info about the call from the VM
+//     exactContextHnd -- [OUT] updated context handle iff call devirtualized
+//
+// Notes:
+//     Virtual calls in IL will always "invoke" the base class method.
+//
+//     This transformation looks for evidence that the type of 'this'
+//     in the call is a final class or would invoke a final method, and
+//     if that and other safety checks pan out, modifies the call and
+//     the call info to create a direct call.
+//
+//     This transformation is done in the importer and not in some
```

Comment:
This paragraph reads like we only call this before inline, but the same method gets called for late devirtualization, right?